causallib/time_varying/GFormula.py

```diff
@@ -0,0 +1,199 @@
+                                    ) -> pd.DataFrame:
+        """
+              Returns individual estimated curves for each subject row in X/a/t
+
+              Args:
+                  X (pd.DataFrame): Baseline covariate matrix of size (num_subjects, num_features).
+                  a (pd.Series): Treatment assignment of size (num_subjects,).
+                  t (pd.Series): Followup durations, size (num_subjects,).
+                  y: NOT USED (for API compatibility only).
+                  timeline_start (int): Common start time-step. If provided, will generate survival curves starting
+                                        from 'timeline_start' for all patients.
+                                        If None, will predict from first observed event (t.min()).
+                  timeline_end (int): Common end time-step. If provided, will generate survival curves up to 'timeline_end'
+                                      for all patients.
+                                      If None, will predict up to last observed event (t.max()).
+
+              Returns:
+                  pd.DataFrame: with time-step index, subject IDs (X.index) as columns and ??
+              """
+        pass
+
+    @abc.abstractmethod
+    def apply_noise(self):
+        pass
+
+    @abc.abstractmethod
+    def prepare_data(self, pat_data):
+        pass
+
+    @staticmethod
+    def _predict_trajectory(self, X, a, t) -> pd.DataFrame:
+        """ Predicts the trajectories for all covariates X and treatment a.
+
+            Args:
+                  X (pd.DataFrame): Baseline covariate matrix of size (num_subjects, num_features).
+                  a (pd.Series): Treatment assignment of size (num_subjects,).
+                  t (pd.Series): Followup durations, size (num_subjects,).
+
+            Returns:
+                 pd.DataFrame: with time-step index, subject IDs (X.index and a) columns and
+                  point values for each column as entries
+        """
+        pass
+
+
+class GFormula(GFormulaBase):
+    """
+        GFormula class that is based on Monte Carlo Simulation for creating noise. #TODO more on this
+    """
+    def __init__(self,
+                 outcome_model,
+                 treatment_model,
+                 covariate_models,
+                 refit_models=True
+                 ):
+
+        super(GFormulaBase).__init__(outcome_model, treatment_model, covariate_models, refit_models)
+
+    def fit(self,
+            X: pd.DataFrame,
+            a: pd.Series,
+            t: pd.Series,
+            y: Optional[Any] = None,
+            refit_models=True,
+            **kwargs
+            ):
+        if refit_models:
+
+            self.treatment_model.fit(df=X, duration_col=t, event_col=y,
+                                    **kwargs)
+            self.treatment_model.fit(X, a)
+
+
+
+            for cov in self.covariate_models:
+                self.covariate_models[cov].fit(X, a)
```

Comment:
For G-computation we may need to a helper to massage the data. the fit should be  on something like

`concat(X, x_1_lagged), a`

Reply:
yes. I believe the entire fitting process is probably more elaborate than that.
I took this PR as just for the basic structure rather than the actual content, no?